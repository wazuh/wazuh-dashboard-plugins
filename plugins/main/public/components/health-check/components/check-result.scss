.wz-health-check-result-disabled {
  opacity: 0.5;
  font-weight: normal;
}
