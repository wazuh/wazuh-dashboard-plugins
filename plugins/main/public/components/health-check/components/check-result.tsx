/*
 * Wazuh app - Check Result Component
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */
import React, { useEffect, useState, useCallback, useMemo, useRef } from 'react';
import {
  EuiDescriptionListDescription,
  EuiDescriptionListTitle,
  EuiCodeBlock,
} from '@elastic/eui';
import InspectLogButton from './inspect-logs-button';
import ResultIcons from './result-icons';


type Result = 'loading' | 'ready' | 'error' | 'error_retry' | 'disabled' | 'waiting';

export function CheckResult(props) {
  const [result, setResult] = useState<Result>('waiting');
  const [isCheckStarted, setIsCheckStarted] = useState<boolean>(false);
  const [verboseInfo, setVerboseInfo] = useState<string[]>([]);
  const [verboseIsOpen, setVerboseIsOpen] = useState<boolean>(false);
  const [isCheckFinished, setIsCheckFinished] = useState<boolean>(false);
  const verboseInfoWithNotificationWasOpened = useRef(false);

  useEffect(() => {
    if (props.shouldCheck && !props.isLoading && awaitForIsReady()){
      initCheck();
    } else if (props.shouldCheck === false && !props.checksReady[props.name]) {
      setResult('disabled');
      setAsReady();
    }
  }, [props.shouldCheck, props.isLoading, props.checksReady]);

  useEffect(() => {
    if (isCheckFinished){
      const errors = verboseInfo.filter(log => log.type === 'error');
      if(errors.length){
        props.canRetry ? setResult('error_retry') : setResult('error');
        props.handleErrors(props.name, errors.map(({message}) => message));
      }else{
        setResult('ready');
        setAsReady();
      }
    }
  }, [isCheckFinished])


  const setAsReady = () => {
    props.handleCheckReady(props.name, true);
  };

  /**
   * validate if the current check is not started and if the depending checks are ready
   */
  const awaitForIsReady = () => {
    return !isCheckStarted && (props.awaitFor.length === 0 || props.awaitFor.every((check) => {
      return props.checksReady[check];
    }))
  }

  const checkLogger = useMemo(() => ({
    _log: (message: string, type: 'info' | 'error' | 'action' ) => {
      setVerboseInfo(state => [...state, {message, type}]);
    },
    info: (message: string) => checkLogger._log(message, 'info'),
    error: (message: string) => checkLogger._log(message, 'error'),
    action: (message: string) => checkLogger._log(message, 'action')
  }), []);

  const initCheck = async () => {
    setIsCheckStarted(true);
    setResult('loading');
    setVerboseInfo([]);
    props.cleanErrors(props.name);
    setIsCheckFinished(false);
    try {
      await props.validationService(checkLogger);
    } catch (error) {
      checkLogger.error(error.message || error);
    }
    setIsCheckFinished(true);
  };


  const checkDidSomeAction = useMemo(() => verboseInfo.some(log => log.type === 'action'), [verboseInfo]);
  const shouldShowNotification = checkDidSomeAction && !verboseInfoWithNotificationWasOpened.current;
  const haveLogs = verboseInfo.length > 0;

  const switchVerboseDetails = useCallback(() => {
    if(checkDidSomeAction && !verboseInfoWithNotificationWasOpened.current){
      verboseInfoWithNotificationWasOpened.current = true;
    };
    setVerboseIsOpen(state => !state);
  },[checkDidSomeAction]);

  const showLogButton = (props.showLogButton && isCheckStarted && haveLogs);

  return (
    <>
      <EuiDescriptionListTitle>
        {props.title}
      </EuiDescriptionListTitle>
      <EuiDescriptionListDescription>
        <p><ResultIcons result={result} initCheck={initCheck}>
        {showLogButton ? <InspectLogButton verboseIsOpen={verboseIsOpen} haveLogs={haveLogs} switchVerboseDetails={switchVerboseDetails} shouldShowNotification={shouldShowNotification} />:<></>}
        </ResultIcons></p>
      </EuiDescriptionListDescription>
      {verboseInfo.length > 0 && (
        <EuiCodeBlock
          className={`wz-width-100 ${verboseIsOpen ? "visible" : ""}`}
          paddingSize='s'
          fontSize='s'
          isCopyable
          style={{textAlign: 'left'}}
        >
          {verboseInfo.map(log => `${log.type.toUpperCase()}: ${log.message}`).join('\n')}
        </EuiCodeBlock>
      )}
    </>
  );
}
