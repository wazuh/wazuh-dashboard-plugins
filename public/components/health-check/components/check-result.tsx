/*
 * Wazuh app - Check Result Component
 *
 * Copyright (C) 2015-2021 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */
import React, { useEffect, useState } from 'react';
import {
  EuiButtonIcon,
  EuiDescriptionListDescription,
  EuiDescriptionListTitle,
  EuiIcon,
  EuiLoadingSpinner,
  EuiToolTip
} from '@elastic/eui';

type Result = 'loading' | 'ready' | 'error' | 'error_retry' | 'disabled' | 'waiting';

export function CheckResult(props) {
  const [result, setResult] = useState<Result>('waiting');
  const [isCheckStarted, setIsCheckStarted] = useState<boolean>(false);

  useEffect(() => {
    if (props.check && !props.isLoading && awaitForIsReady()){
      initCheck();
    } else if (props.check === false && !props.checksReady[props.name]) {
      setResult('disabled');
      setAsReady();
    }
  }, [props.check, props.isLoading, props.checksReady]);

  const setAsReady = () => {
    props.handleCheckReady(props.name, true);
  };

  const handleErrors = (errors, parsed?) => {
    props.handleErrors(props.name, errors, parsed);
  };

  /**
   * validate if the current check is not started and if the dependentes checks are ready
   */
  const awaitForIsReady = () => {
    return !isCheckStarted && (props.awaitFor.length === 0 || props.awaitFor.every((check) => {
      return props.checksReady[check];
    }))
  }

  const initCheck = async () => {
    setIsCheckStarted(true);
    setResult('loading');
    props.cleanErrors(props.name);
    try {
      const { errors } = await props.validationService();
      if (errors.length) {
        handleErrors(errors);
        setResult('error');
        props.canRetry ? setResult('error_retry') : setResult('error');
      } else {
        setResult('ready');
        setAsReady();
      }      
    } catch (error) {
      handleErrors([error], true);
      props.canRetry ? setResult('error_retry') : setResult('error');
    }
  };

  const renderResult = () => {
    switch (result) {
      case 'loading':
        return (
          <span>
            <EuiLoadingSpinner size="m" /> Checking...
          </span>
        );
      case 'disabled':
        return <span>Disabled</span>;
      case 'ready':
        return (
          <span>
            <EuiIcon type="check" color="secondary"></EuiIcon> Ready
          </span>
        );
      case 'error':
        return (
          <span>
            <EuiIcon type="alert" color="danger"></EuiIcon> Error
          </span>
        );
      case 'error_retry':
        return (
          <span>
            <EuiIcon type="alert" color="danger"></EuiIcon> Error
            <EuiToolTip
              position='top'
              content='Retry'
            >
              <EuiButtonIcon
                display="base"
                iconType="refresh"
                iconSize="m"
                onClick={initCheck}
                size="m"
                aria-label="Next"
              />
            </EuiToolTip>
          </span>
        );
      case 'waiting':
        return (
          <span>
            Waiting...
          </span>
        );
    }
  };

  return (
    <>
      <EuiDescriptionListTitle>{props.title}</EuiDescriptionListTitle>
      <EuiDescriptionListDescription>{renderResult()}</EuiDescriptionListDescription>
    </>
  );
}
