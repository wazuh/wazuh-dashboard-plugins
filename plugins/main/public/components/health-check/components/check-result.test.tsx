/*
 * Wazuh app - Check Result Component - Test
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */
import React from 'react';
import { mount } from 'enzyme';
import { CheckResult } from './check-result';
import { act } from 'react-dom/test-utils';

describe('Check result component', () => {
  const validationService = jest.fn();
  const handleErrors = jest.fn();
  const handleCheckReady = jest.fn();
  const cleanErrors = jest.fn();

  const awaitForMyComponent = async (wrapper: any) => {
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
      wrapper.update();
    });
  };

  it('should render a Check result screen', async () => {
    validationService.mockImplementation(() => ({ errors: [] }));
    const component = mount(
      <CheckResult
        name={'test'}
        title={'Check Test'}
        awaitFor={[]}
        check={true}
        validationService={validationService}
        handleErrors={handleErrors}
        isLoading={false}
        handleCheckReady={handleCheckReady}
        checksReady={{}}
        cleanErrors={cleanErrors}
        canRetry={true}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should print ready', async () => {
    validationService.mockImplementation(() => ({ errors: [] }));
    const wrapper = await mount(
      <CheckResult
        name={'test'}
        title={'Check Test'}
        awaitFor={[]}
        shouldCheck={true}
        validationService={validationService}
        handleErrors={handleErrors}
        isLoading={false}
        handleCheckReady={handleCheckReady}
        checksReady={{}}
        cleanErrors={cleanErrors}
        canRetry={true}
      />
    );

    await awaitForMyComponent(wrapper);

    expect(wrapper.find('ResultIcons').exists()).toBeTruthy();
    expect(wrapper.find('ResultIcons').prop('result')).toBe('ready');
  });

  it('should print error_retry', async () => {
    validationService.mockImplementation(() => {
      throw 'error_retry';
    });
    const wrapper = await mount(
      <CheckResult
        name={'test'}
        title={'Check Test'}
        awaitFor={[]}
        shouldCheck={true}
        validationService={validationService}
        handleErrors={handleErrors}
        isLoading={false}
        handleCheckReady={handleCheckReady}
        checksReady={{}}
        cleanErrors={cleanErrors}
        canRetry={true}
      />
    );

    await awaitForMyComponent(wrapper);

    expect(wrapper.find('ResultIcons').exists()).toBeTruthy();
    expect(wrapper.find('ResultIcons').prop('result')).toBe('error_retry');
  });

  it('should print error', async () => {
    validationService.mockImplementation(() => {
      throw 'error';
    });
    const wrapper = await mount(
      <CheckResult
        name={'test'}
        title={'Check Test'}
        awaitFor={[]}
        shouldCheck={true}
        validationService={validationService}
        handleErrors={handleErrors}
        isLoading={false}
        handleCheckReady={handleCheckReady}
        checksReady={{}}
        cleanErrors={cleanErrors}
        canRetry={false}
      />
    );

    expect(wrapper.find('ResultIcons').exists()).toBeTruthy();
    expect(wrapper.find('ResultIcons').prop('result')).toBe('error');
  });
});
