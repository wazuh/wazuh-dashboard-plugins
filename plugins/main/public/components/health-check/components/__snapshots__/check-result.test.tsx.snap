// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check result component should render a Check result screen 1`] = `
<CheckResult
  awaitFor={Array []}
  canRetry={true}
  check={true}
  checksReady={Object {}}
  cleanErrors={[MockFunction]}
  handleCheckReady={[MockFunction]}
  handleErrors={[MockFunction]}
  isLoading={false}
  name="test"
  title="Check Test"
  validationService={[MockFunction]}
>
  <EuiDescriptionListTitle>
    <dt
      className="euiDescriptionList__title"
    >
      Check Test
    </dt>
  </EuiDescriptionListTitle>
  <EuiDescriptionListDescription>
    <dd
      className="euiDescriptionList__description"
    >
      <p>
        <ResultIcons
          initCheck={[Function]}
          result="waiting"
        >
          <EuiToolTip
            content="On hold..."
            delay="regular"
            position="top"
          >
            <span
              className="euiToolTipAnchor"
              onKeyUp={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <EuiIcon
                aria-label="waiting"
                color="#999999"
                onBlur={[Function]}
                onFocus={[Function]}
                type="clock"
              >
                <EuiIconEmpty
                  aria-hidden={true}
                  aria-label="waiting"
                  className="euiIcon euiIcon--medium euiIcon-isLoading"
                  focusable="false"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  role="img"
                  style={
                    Object {
                      "fill": "#999999",
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    aria-label="waiting"
                    className="euiIcon euiIcon--medium euiIcon-isLoading"
                    focusable="false"
                    height={16}
                    onBlur={[Function]}
                    onFocus={[Function]}
                    role="img"
                    style={
                      Object {
                        "fill": "#999999",
                      }
                    }
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </EuiIconEmpty>
              </EuiIcon>
            </span>
          </EuiToolTip>
        </ResultIcons>
      </p>
    </dd>
  </EuiDescriptionListDescription>
</CheckResult>
`;
